### PR TITLE
[HACK] Make concretization great again!

### PR DESCRIPTION
**Pull request by adamjstewart**
I _really_ didn't want to have to do this, but I've been left with no other choice. Spack's concretization algorithm is so crippled at the moment that simple commands no longer work more often than I would like. This PR does not fix the concretizer, as I wouldn't even no where to begin.

So what does this PR do?

- [x] Always default to `+mpi` and `~X` for _every_ package
- [x] Never even try to forward variants, as it doesn't work anyway

Why in the world would you want to do something so drastic?

Try checking out a fresh copy of Spack, and with no `packages.yaml`, run the following command:
```
$ spack spec gtkplus
...
==> Error: Invalid spec: 'pango@1.40.1%gcc@6.2.1~X arch=linux-fedora25-x86_64'. Package pango requires variant +X, but spec asked for ~X
```
The problem here is that `gtkplus` and `cairo` default to `+X` while `pango` defaults to `~X`. Spack's concretization algorithm is incapable of resolving this conflict because it settles on the default value for all dependencies. This bug has been reported over and over and over again (see #2546, #2574, #1552, etc.).

The solution that I decided on in this PR is to make sure common variants like `+X` and `+mpi` are at least consistent. Of course, as soon as you run `spack spec gtkplus+X`, everything breaks again, but at least `spack spec gtkplus` works.

What solutions have we used in the past?

When going through the packages, I noticed a lot of packages that looked like:
```python
depends_on('hdf5+mpi', when='+mpi')
depends_on('hdf5~mpi', when='~mpi')
```
What's wrong with this? Well, first of all, it doesn't forward the variants properly, as seen with `spack spec gtkplus`. Second, try running the following:
```
spack spec packageA ^hdf5@1.8.12
```
With a packageA containing those two lines, it will inform you that packageA does not depend on hdf5. This is because Spack doesn't concretize default variants before it concretizes dependency versions. Either of the following commands will work:
```
spack spec packageA +mpi ^hdf5@1.8.12
spack spec packageA -mpi ^hdf5@1.8.12
```
but the original won't. So what if we always do:
```python
depends_on('hdf5')
depends_on('hdf5+mpi', when='+mpi')
depends_on('hdf5~mpi', when='~mpi')
```
Ok, first of all, that's a pain in the ass. Second of all, it doesn't work, as evidenced by `spack spec gtkplus`. But what if we did:
```python
depends_on('hdf5')
depends_on('hdf5+mpi', when='+mpi')
```
Doesn't work either. I just tried it with the `gtkplus`, `cairo`, `pango` chain and you still see the problem with a conflict between `+X` and `~X`.

But won't everything be happy if we keep variant forwarding and make sure all of the variants have the same default? Nope. Things like `spack spec gtkplus` will work, but `spack spec gtkplus +X` won't. And if you decide to change the default for any package in `packages.yaml`, we're right back to square one, except it's harder to debug now unless you tell us what is in your `packages.yaml`.

There is another standing issue that this PR does not address. Try running the following:
```
spack spec python+tk
...
==> Error: +tk does not satisfy ~tk
```
This may look like the problem we just solved, but it's not that simple. Here is the dependency chain:

python+tk -> tk -> libx11 -> libxcb -> python~tk

Since libxcb has a build dependency on python~tk, it says it doesn't satisfy python+tk. Of course, if you try to fix this by setting the default to `python+tk`, you end up in infinite recursion hell, as seen in #2495 and #2565. Even more fun, if you try to build py-matplotlib for Python 3, it requires python+tk, and you end up with the problem that xcb-proto, another dependency of libxcb, depends on Python 2, which is most certainly not Python 3. 

This will hopefully be fixed when we start resolving build dependencies separately, but for now it is an absolute nightmare.

I _really_ hope this PR won't be merged, but it does provide immediate relief for a lot of open issues. In my opinion, these issues should block the release of `0.10`. If users can't build `python+tk` or `gtkplus` without hacking up Spack, we have a real problem.

### CODE REVIEW

**Comment by adamjstewart** [comment]
If this PR were merged, how would you forward variants?

Well, for the case of gtkplus, you would have to either run:
```
spack install gtkplus+X ^cairo+X ^pango+X
```
or set them in your `packages.yaml`. But at least it wouldn't crash! The danger then becomes there is nothing stopping someone from running:
```
spack install gtkplus+X
```
and later discovering that it doesn't work because the dependencies were built without X support. The stance I took to address this in #1553 was to raise an error if `gtkplus+X` is being built with `cairo~X` or `pango~X`. This is obviously not ideal.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -72,20 +72,7 @@ class Dealii(CMakePackage):
...
-    depends_on("boost@1.59.0:+thread+system+serialization+iostreams",
-               when='@8.5.0:~mpi~python')
-    depends_on("boost@1.59.0:+thread+system+serialization+iostreams+mpi",
-               when='@8.5.0:+mpi~python')
-    depends_on("boost@1.59.0:+thread+system+serialization+iostreams+python",
-               when='@8.5.0:~mpi+python')
-    depends_on(
-        "boost@1.59.0:+thread+system+serialization+iostreams+mpi+python",
-        when='@8.5.0:+mpi+python')
+    depends_on("boost@1.59.0:+thread+system+serialization+iostreams")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: Frankly, i don't understand why you do this. You say

> Well, first of all, it doesn't forward the variants properly

but in my experience things work perfectly fine in this regards.

I agree that if you would try to specify which `boost` is needed here, you would need to do `spack spec dealii+mpi+python ^boost@1.60.0` or alike, it's PITA but if you put your preferences to `packages.yaml`, then all works fine, at least from my experience.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ class Petsc(Package):
     depends_on('boost', when='@:3.5+boost')
     depends_on('metis@5:', when='+metis')
 
-    depends_on('hdf5+mpi', when='+hdf5+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: this is not even different from the original, before it will use `hdf5` if-and-only-if `petsc` is compiled with `mpi`. So it's not even about forwarding of variants. Same below.

**Comment by eschnett** [comment]
In general, why would a package explicitly require `otherpackage~feature`, _unless that package really doesn't build in that case_? It seems that half the confusion that this PR addresses is that some packages say `hdf5~mpi` when they mean "I need HDF5, but I don't need MPI". That's wrong, as `hdf5~mpi` means "I need HDF5, and I really cannot handle an HDF5 with MPI at all". Grepping for the character `~` in dependencies should find most of these; one would assume that there's always a comment explaining the situation. (I assume that some cases were added merely out of desperation to make concretization work.)

Idea: Add a new named argument `why_not` to `depends`, and require it to be set to an explanatory string when the `~` character is used to exclude a feature from a dependency...

Finally, `mpi` is a special case since if there's MPI anywhere in the dependency chain, there's a good chance that `cc` will fail and `mpicc` needs to be used. I say that's a bug in the MPI package; it should probably change Spack's `cc` setting to its `mpicc`, or -- even better -- should declare its include and library requirements in a transparent manner instead of pointing to an opaque `mpicc` script. (I know that I volunteered to fix this, but sadly didn't have the time to do so yet.)

**Comment by davydden** [comment]
if there are any changes to contretizer coming from @tgamblin , i would not merge this PR until they are merged upstream and the problem is reassessed.

**Comment by davydden** [comment]
@eschnett :

>  if there's MPI anywhere in the dependency chain, there's a good chance that cc will fail and mpicc needs to be used.

Can you please elaborate? It should be perfectly fine to compile a serial-only `packageA` with GCC and then use it in another `packagesB` that utilize some MPI (say Openmpi compiled with GCC).

**Comment by citibeth** [comment]
@adamjstewart Thank you for writing this up and bringing the issues to our attention.  My thoughts are:

1. The write-up is valuable.  But as @davydden said, it feels a bit all over the place.  It's clear that things are broken and that this is frustrating, but it's not clear yet exactly what is broken or how we think it should be fixed.  I think some editing of it would do us all service: to try to be more specific about WHAT the problems are, WHAT doesn't work, and HOW those things don't work.  Maybe I could take some time with what you've written and a bunch of `spack spec` tries to flesh this out?

2. I'm wondering why you don't think this PR should be merged, since it looks to me like it does some good things.  Namely:
  a) Variants should generally default the same way in different packages.
  b) As you point out, the ubiquitous attempts at manual variant forwarding are not working --- and they introduce at least as many problems as they solve.  I think that getting rid of them is a good start.  Clearly, we need to finish the job with Spack-supported variant forwarding or universal variants.  But either way, we need to clear out the manual attempts.

**Comment by citibeth** [comment]
BTW... I've finally articulated how the multi-valued variants in #2386 are different from boolean variants; and this could guide us in deciding when to use which feature.  With boolean variants, a request for `pkg` will match `pkg+x`.  But with multi-valued variants, I presume that a request for `pkg-x=y` will only match `pkg-x=y`.

File: var/spack/repos/builtin/packages/gmsh/package.py [hunk 3]
```diff
@@ -66,8 +66,7 @@ class Gmsh(Package):
     depends_on('fltk', when='+fltk')
     depends_on('hdf5', when='+hdf5')
     depends_on('oce',  when='+oce')
-    depends_on('petsc+mpi', when='+petsc+mpi')
-    depends_on('petsc', when='+petsc~mpi')
+    depends_on('petsc', when='+petsc')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/gmsh/package.py]: here you allow to build `gmsh+mpi` against `petsc~mpi`, I don't think it's a correct behaviour.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 4]
```diff
@@ -96,20 +83,20 @@ class Dealii(CMakePackage):
 
     # optional dependencies
     depends_on("mpi",              when="+mpi")
-    depends_on("arpack-ng+mpi",    when='+arpack+mpi')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dealii/package.py]: I was confused about this as well. So, according to the way it used to be, if I build `+arpack~mpi` it doesn't build with arpack, only when I build with `+arpack+mpi`? That doesn't seem... ideal. What if we raised an error when someone tries to build `+arpack~mpi`?

In hindsight, this wasn't the change I should have made. I should have removed `arpack-ng+mpi`, but not `+arpack+mpi`.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py]: Yeah, you're right. I probably shouldn't have made this change.

**Comment by eschnett** [comment]
@davydden Let me take HDF5 as example. If configured with `+mpi`, the HDF5 header files contain an `#include <mpi.h>`, and (obviously) also require the MPI libraries for linking. If you build another serial package against this HDF5 version, then you need to build and link it (although it is serial!) with `mpicc`.

If the MPI package properly declared its include and library requirements to Spack, then Spack would handle this automatically.

**Comment by tgamblin** [comment]
Hi all: just to clarify, the issue is really that the concretizer does not currently do backtracking for decisions made about variants.  It can currently do *one* level of backtracking when it tries to figure out which implementation of a virtual dependency to choose.  What it *should* do is backtrack from much deeper levels as it searches for a satisfiable configuration.  The variants should just be another generalized choice in the solver.

Implementing a fix for this **is** on my list, and I am kind of thinking (based on all the variant conflicts) that it should get moved up in priority.  The dependencies in Spack have gotten more complex now that all of X11 and more fancy MPI options are in Spack.  So I guess people discovered that the dependency model really does model most of the things they want, but the concretizer needs to grow up.

I think a reasonable plan is:

1. Finish the stuff left for `0.10` and get it out the door.
2. I already pushed the exact versions stuff from `0.10` off to `1.0` because the changes to the concretizer were more sweeping that I would've liked. There are some concepts that have become conflated over time in the constraint solve ("satisfies" vs. "could satisfy") that need refactoring, and I don't trust the refactor without better package testing.
3. This change should probably be done concurrently with (2), as it affects the main part of the concretization algorithm.

So my suggestion is that we put this in the `1.0` bin and I can focus on it with @mathstuf @becker33 and @scheibelp between now and January-ish.

Thoughts?

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 4 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: >  if I build +arpack~mpi it doesn't build with arpack, only when I build with +arpack+mpi? That doesn't seem... ideal

correct, that's the behaviour and it is indeed not ideal, but variant description clearly say so: `(only with MPI)`.

In this specific case of `arpack`, one actually can build (outside of Spack) `dealii+mpi^arpack~mpi`. Here it was only added for consistency. But other packages may get broken when you try this (it all depends).

**Comment by alalazo** [comment]
> There are some concepts that have become conflated over time in the constraint solve ("satisfies" vs. "could satisfy") that need refactoring, **and I don't trust the refactor without better package testing.**

This is worth another :+1:

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 5]
```diff
@@ -96,20 +83,20 @@ class Dealii(CMakePackage):
...
     depends_on("gsl",              when='@8.5.0:+gsl')
-    depends_on("hdf5+mpi",         when='+hdf5+mpi')
+    depends_on("hdf5",             when='+hdf5')
     depends_on("metis@5:",         when='+metis')
-    depends_on("netcdf+mpi",       when="+netcdf+mpi")
-    depends_on("netcdf-cxx",       when='+netcdf+mpi')
+    depends_on("netcdf",           when="+netcdf")
+    depends_on("netcdf-cxx",       when='+netcdf')
     depends_on("oce",              when='+oce')
-    depends_on("p4est",            when='+p4est+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: this is for example wrong, `p4est` builds only with `mpi`, so in this instance `deal.ii` can use `p4est` if and only if it is build with `mpi` itself. I would certainly recommend splitting this PR into at least

1. `mpi=True` by default
2. the rest...

So that one could take a closer look. I am afraid you may be breaking more things that you try to fix here.

**Comment by tgamblin** [comment]
More thoughts: to mirror what other said, it's very helpful that @adamjstewart has written this up.  The changes in this PR seem like they are intended to hard-code the best "common-case" choices for things, so that for unconstrained specs, things work decently and so that had package constraints don't throw too many monkey wrenches in the path to satisfiability.

We *could* merge this as a stop gap, but there is useful information in those constraints.  I am not completely convinced that this PR is going to make things work better for most people.  I think I'd need tests to see that.

At a really high level, my immediate priorities right now are:

1. Get the last few bugs out the door for `0.10`
1. Getting a CDash site set up with a reasonably comprehensive set of package builds so that we can see what works and what does not.  Right now we are playing whack-a-mole with github issues and we need a dashboard so we can make the thing converge to green.
1. Get the concretizer fixes in.

Does that sound like a reasonable set of priorities?  I think (2) is necessary before we can do (3) right.  Also, getting (2) working means we can start recording many of the tests cases @adamjstewart mentioned above in a set of nightly package tests.  This makes (3) work that much better.

**Comment by adamjstewart** [comment]
> frankly i don't understand what's dramatically wrong with
> ```python
> depends_on('hdf5+mpi`, when='+mpi')
> depends_on('hdf5~mpi`, when='~mpi')
> ```

Let's take a quick walkthrough with `gtkplus`, as it is the simplest problem to understand. Currently in develop, we have this set of conditions:

**gtkplus**
```python
variant('X', default=True)

depends_on("pango")                                                         
depends_on("pango~X", when='~X')                                            
depends_on("pango+X", when='+X')
```
**pango**
```python
variant('X', default=False)

depends_on("cairo")                                                         
depends_on("cairo~X", when='~X')                                            
depends_on("cairo+X", when='+X')
```
**cairo**
```python
variant('X', default=True)
```
If you run `spack spec gtkplus` with no `packages.yaml`, you get:
```
==> Error: Invalid spec: 'pango@1.40.1%gcc@6.2.1~X arch=linux-fedora25-x86_64'. Package pango requires variant +X, but spec asked for ~X
```
If things were working properly, the concretizer would notice that I didn't specify any variants, default to `gtkplus+X`, and propagate `+X` down the line with the `depends_on` statements. But it isn't.

What are some possible solutions? What if we set the `~X` to be the default for all 3 packages? Now, the following commands work:
```
spack spec gtkplus
spack spec gtkplus +X
```
But the following doesn't work:
```
spack spec gtkplus ^pango+X
```
In an ideal world, Spack would say that since pango needs `+X`, so does `gtkplus`, but it doesn't work. What about another scenario. If we set `~X` to be the default for all packages, and a user happens to always need `cairo+X`, they could modify their `packages.yaml` like so:
```yaml
packages:
  cairo:
    variants: +X
```
That's perfectly legal right? But now `spack spec gtkplus` no longer works.

**Comment by luigi-calori** [comment]
I' m feeling really sorry as it seems I ( unintentionally ) opened a vase of pandora.

I can summarize here my intended use for spack and hopefully get some hints:

I' m trying to use spack to automate the building of a sw stack for remote visualization.
One of the target for installation is a cluster that has no GPU available, so one of the goal was to build Paraview with mesa.

Ideally this could be handled by a single spack install command..... but I do not expect it would be so easy.

I raised the issue and the following PR thinking to just fix typos in some underneath packages.

Unfortunately  problems I detected involve spack internals such as concretization, and could not be solved at the package level.
Nevertheless, as I would really like to be able to code a working recipy in spack, any suggestion to good practice that relieve these problems would be really welcomed.

Many of you suggest using package.yaml:  would it be possible to collect some real examples ( if it can be openly disclosed) of configuration that are really in use?

As some build steps ( as qt ) are really long, even splitting install into steps ( build tools, base libraries, applications ) could be acceptable, providing some hints on how to properly set up package.yaml to avoid rebuilding dependencies in following step.

**Comment by citibeth** [comment]
Thank you for the detailed example.  I question one thing in it:

>  In an ideal world, Spack would say that since pango needs +X, so does
> gtkplus,
>
Let's think in terms of PATTERNS and SPECS.  They both have the same
syntax, so we tend to get them confused.

I thought the way things were SUPPOSED to work is that the PATTERN
`packageA` will match the SPEC `packageA+x` or `packageA~x`.  The rationale
is that if `packageB` -> `packageA` without anything else, then `packageB`
should not really care whether or not `packageA` does or does not have the
`+x` feature turned on.  Boolean variants really do mean that "this is an
extra feature."  If not, then maybe multi-valued variants should be used
instead (once they're merged).

So... I believe that THIS little corner works as it should for now: a non-x
gtk should work find with pango+x

**Comment by citibeth** [comment]
Luigi, don't be sorry.  This Pandora's box was already open, and we're
still grasping at what to do about it.

> Many of you suggest using package.yaml: would it be possible to collect
some real examples

I use point to system-supplied Qt, Python2 and any X libraries in my
`packages.yaml`.  This solves a lot of problems (unless your system doesn't
have them).

> As some build steps ( as qt ) are really long

I use variants to turn off package options that use Qt --- except for
`py-matplotlib`, where I actually do care about a GUI.

**Comment by davydden** [comment]
@adamjstewart 

> But the following doesn't work:
> spack spec gtkplus ^pango+X
> In an ideal world, Spack would say that since pango needs +X, so does gtkplus, but it doesn't work. 

thanks for explanation, now i see the problem. 

I would still suggest do it step-by-step. First part of your PR which makes variants consistently `true` by default. This can go into `0.10` as it is won't break things. The rest needs to be tested, preferably grouped into suites, say all PDE packages like `petsc`, `deal.ii`, `p4est`, `fenics`, `gmsh`.

**Comment by davydden** [comment]
@adamjstewart : 

> But the following doesn't work:
> spack spec gtkplus ^pango+X

it it because you have 
```
depends_on("pango")                        
```
in addition to
```                                 
depends_on("pango~X", when='~X')                                            
depends_on("pango+X", when='+X')
```
? The last two should be sufficient. Maybe this is the root of the bug?

**Comment by citibeth** [comment]
@davydden  I agree with @adamjstewart that we should not be doing stuff like :
```
depends_on("pango~X", when='~X')                                            
depends_on("pango+X", when='+X')
```
It breaks the principle I mentioned above about how patterns are matched to specs.

I don't think these problems will be solved until we agree on a scheme for variant forwarding / universal variants.

**Comment by davydden** [comment]
@eschnett 

> @davydden Let me take HDF5 as example. If configured with +mpi, the HDF5 header files contain an #include <mpi.h>, and (obviously) also require the MPI libraries for linking. If you build another serial package against this HDF5 version, then you need to build and link it (although it is serial!) with mpicc.

that's another direction (use mpi-build package in non-mpi package), but I agree with this, of course. I would say that' exactly we have 
```
depends_on(package+mpi, when='+mpi')
depends_on(package~mpi, when='~mpi`)
```

But similar issues arise not only with `mpi`. Example [above] is 32/64 bit integers where some dependencies may need to be turned off if they don't support 64bit ints.  Same thing with complex/real valued, for example if you build `petsc+complex`, you can't use `hypre` preconditioner. 

Bottom line is: I don't think we need to make some exceptions for `mpi`, those constraints arise in other context as well. So we (as a community) just need to have conretizer which is robust to handle such cases.

**Comment by davydden** [comment]
@citibeth 

> @davydden I agree with @adamjstewart that we should not be doing stuff like :

I respectfully disagree, i think it's a valid deterministic approach to specify constraints. Variant forwarding will only solve `mpi`-related issues, but will not cover other similar usage cases, see my [previous post](https://github.com/LLNL/spack/pull/2590#issuecomment-267139972).

**Comment by davydden** [comment]
@adamjstewart :

> But the following doesn't work:
> spack spec gtkplus ^pango+X

going to this explanation, one would just need to help spack with `spack spec gtkplus+X ^pnago+X`, right? I think it is an OK price to pay for now when there are cases like
```
depends_on(packageA+X, when='+Y')
depends_on(packageA~X, when='~Y`)
```
irrespectively of variant forwarding as this `Y` may be different from `X` (but related). So there could be some extra things like
```
depends_on(packageC, when='+Y')
depends_on(packageD, when='~Y')
```
I think it's not as bad, just inconvenient at times.

**Comment by citibeth** [comment]
On Wed, Dec 14, 2016 at 3:00 PM, Denis Davydov <notifications@github.com>
wrote:

> @eschnett <https://github.com/eschnett>
>
> @davydden <https://github.com/davydden> Let me take HDF5 as example. If
> configured with +mpi, the HDF5 header files contain an #include <mpi.h>,
> and (obviously) also require the MPI libraries for linking. If you build
> another serial package against this HDF5 version, then you need to build
> and link it (although it is serial!) with mpicc.
>
>
The linking is taken care of automatically, no need to explicitly link to
MPI.

**Comment by citibeth** [comment]
> I respectfully disagree, i think it's a valid deterministic approach to
> specify constraints. Variant forwarding will only solve mpi-related
> issues, but will not cover other similar usage cases, see my previous post
> <https://github.com/LLNL/spack/pull/2590#issuecomment-267139972>.
>

@tgamblin Then we will need to change a fundamental way that
pattern-matching currently works.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 5 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dealii/package.py]: I did separate things into 3 different commits for this reason. It sounds like the first 2 commits are popular, but the 3rd is much more controversial.

Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: > I did separate things into 3 different commits for this reason.

please cherry-pick the first one into a separate PR, i think it can be merged almost immediately (but i would certainly like to review it). Then the second one (same here). After that this PR can be rebased and it would be easier to discuss.

**Comment by tgamblin** [comment]
@citibeth: 

> @tgamblin Then we will need to change a fundamental way that pattern-matching currently works.

No, we won't.  This doesn't violate anything.

Let `->` mean `depends_on`.

Consider :
* `A -> B` means `A` requires `B` but doesn't care how it is built.  `A+mpi` or `A~mpi` will do.
* `A -> B+mpi` means that the `B` used to satisfy `A` must have `+mpi`.  
* `A -> B~mpi` means that the `B` used to satisfy `A` must have `~mpi`.  

These are all valid builds if the only constraint is `A -> B`:

* `A+mpi ^B+mpi`
* `A~mpi ^B+mpi`
* `A+mpi ^B~mpi`
* `A~mpi ^B~mpi`

Adding these two constraints to a package:
* `A -> B+mpi when A+mpi`
* `A -> B~mpi when A~mpi`

Just means that these are the only two valid builds:

* `A+mpi ^B+mpi`
* `A~mpi ^B~mpi`

Any concrete `A` MUST have a value for the `mpi` variant, so `A`, concretized, will become either `A+mpi` or `A~mpi`.  There is no "don't care" build -- it either has MPI or it doesn't when it's actually built.  The constraints you're arguing against just impose a requirement on `B` based on how `A` is built and don't allow builds where one has MPI and the other doesn't.

What is the matter with that?

**Comment by adamjstewart** [comment]
It sounds like everyone is on board with the first 2 commits that make sure that all `mpi` and `X` variants have the same default. Are there any other common variants that should be aligned? Of course, things quickly break if you change any of the defaults in `packages.yaml`, but there doesn't seem to be anything we can do about that until the underlying issue (the concretizer) is fixed.

It sounds like a lot of people are not on board with my last commit, which removes the depends_on variant forwarding. I'm fine with removing this commit for now if it helps get the first 2 commits in. I'm not sure how useful those depends_on statements are though. Even if it did work, it is kind of annoying to have to always say:
```python
depends_on('hdf5')
depends_on('hdf5+mpi', when='+mpi')
depends_on('hdf5~mpi', when='~mpi')
```

I think there are two long term solutions. The universal variants/variant forwarding idea that @citibeth and I have proposed would solve the problem of having to be so annoyingly explicit on every dependency, and I think that is worth doing. As for which we go with (universal variants vs. variant forwarding) there are enough open issues discussing the pros and cons of both already. It can wait until after `0.10` I guess.

But the above syntax should still work. That is a bug in the concretizer that @tgamblin explained. That should also be fixed, and I'm ok with waiting until after `0.10` I guess. The bug has always been present. It's just been exacerbated now that software stacks are getting deeper and X11 support has been proliferated.

How about I remove the last commit and we merge the first 2 commits? It's a useful stopgap, although it doesn't solve every situation.

**Comment by davydden** [comment]
> How about I remove the last commit and we merge the first 2 commits? It's a useful stopgap, although it doesn't solve every situation.

agreed.

**Comment by alalazo** [comment]
> How about I remove the last commit and we merge the first 2 commits? It's a useful stopgap, although it doesn't solve every situation.

:+1: 

> As for which we go with (universal variants vs. variant forwarding) there are enough open issues discussing the pros and cons of both already.

The two concepts don't cover the same terrain. Having something "universal" means one single object to represent a property in the entire DAG. It could be useful for things like `shared` and `static` (even though you will lose the opportunity to have them mixed) or other properties that should be uniform across any given build. 

Variant forwarding will help in cases where you need to make a decision based on somebody else. The most relevant example is `openmp`: if multiple packages in a DAG can be OpenMP multi-threaded what you usually want is to activate only one of those and avoid multiple nested regions.

**Comment by hartzell** [comment]
I'm not in deep enough to have a useful opinion on the details, but will contribute the datapoint that as an end user I'm comfortable with the plan and priorities that @tgamblin described up above (interim fixes and fixing the concretizer the right way in early January, if I understand).

While having to do things like `R@3.3.1 ^pango+X` is magical and a bit annoying, there aren't *that* many new users arriving on a daily basis so just helping them (us...) over the hump as needed and fixing the problem well in a timely manner seems perfectly reasonable.  

Between now and then, "*How much can go wrong?*"

**Comment by citibeth** [comment]
> The constraints you're arguing against just impose a requirement on B
> based on how A is built and don't allow builds where one has MPI and the
> other doesn't.
>
> What is the matter with that?
>
OK so it's not a change in the concretizer.  But it is a change in the way
we've interpreted variants up till now.  In most cases, the `+` means we
have an "added feature."  Package A should almost never say it requires B
*without* an extra feature; after all, why should it care how many extra
features B has, as long as it has the features that A needs?  Are we
imposing these extra constraints to cover for a shortcoming elsewhere?  In
the MPI case... why should A care whether or not B links to MPI behind the
scenes?

BTW... the universal variants I've suggested would have the property I'm
arguing against here:  If you turn on universal variant `mpi`, then it will
be enabled for EVERY package in the DAG that uses it.

**Comment by alalazo** [comment]
> BTW... the universal variants I've suggested would have the property I'm
arguing against here:  If you turn on universal variant `mpi`, then it will
be enabled for EVERY package in the DAG that uses it.

`mpi` as a universal variant is a wrong example: what if I always use serial `arpack-ng` or serial `hdf5` as part of a software, regardless of whether this software was compiled with `+mpi` or `~mpi`?

**Comment by tgamblin** [comment]
> But it is a change in the way we've interpreted variants up till now.

Sure, but this is a matter of convention and interpretation.  For the most part, we try to model variants this way, but I don't think that even modeling all variants as `+feature` results in what you're asking for.  Consider @alalazo's OpenMP case.  Some packages can build with OpenMP.  That is a feature. You may need `A+openmp -> B~openmp` to enforce mutual exclusion and to avoid nested OpenMP regions, because OpenMP in `B` actually *conflicts* with OpenMP in `A`.

> In the MPI case... why should A care whether or not B links to MPI behind the scenes?

In most cases, if `B -> mpi` and `A -> B` but `A !-> mpi` , `A` wasn't intended to be run with `mpirun`.  If `A` isn't run with `mpirun`, and some routine from `B` calls something that calls an MPI routine, things break because the code isn't run in parallel.  You'll likely just get an error that some MPI routine was called before `MPI_Init()`.  I *believe* this is the case with HDF5.  You can have an MPI program that links either the serial or parallel HDF5 library but I do not believe you can link a serial program to `hdf5+mpi` and expect it to run correctly.  They may have fixed this in more recent HDF5 versions, but this is always the annoyance with HDF5 libs.  So depending on `hdf5~mpi` is basically depending on serial HDF5.

So basically I want a **logically complete** dependency system, because that lets me express cases like these when I need to.  Figuring out what the best usage conventions are on top of that depends on what's elegant, what works, and what we can do with the current state of HPC software, which has all kinds of nasty linking issues.

**Comment by hartzell** [comment]
And speaking of NP complete problems, references to [this writeup about package managers and SAT](https://research.swtch.com/version-sat) came across the Hacker News radar recently.

**Comment by tgamblin** [comment]
@hartzell: indeed. You're the second person to send me that today.  What we're talking about here is making Spack's concretizer solve basically that problem, but with variants, compilers, and optional dependencies within each build.  I would say the current concretizer is somewhere between heuristics and a full backtracking solver.

The nice thing about Spack is we don't have to solve version SAT for *everything* that is installed at once, just within each build.  Also Spack has some leeway that most PM's don't w.r.t. build dependencies (which can be RPATH'd to versions of libs separate from those in the DAG to be built).

**Comment by davydden** [comment]
@adamjstewart remove the last commit and we can finish with this one?

**Comment by citibeth** [comment]
Git has a revert command that will apply a new commit reversing an existing
commit.

**Comment by davydden** [comment]
> Git has a revert command that will apply a new commit reversing an existing
commit.

IMHO it is easier by doing `git rebase -i HEAD~3` and then removing the last one, and `git push`. My perception is that @adamjstewart separated commits intentionally, so I think one should not squash them during the merge and thereby the history should be clean.

**Comment by adamjstewart** [comment]
Ok, last commit removed. Should be good to go now.
